tfhe/src/core_crypto/algorithms/lwe_encryption.rs

```diff
@@ -1964,67 +2270,73 @@ pub fn encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
+                        mask_noise,
+                    ),
+                    body_noise,
                 ),
             )| {
-                // output_body_chunk may not be able to fit the full convolution result so we create
-                // a temp buffer to compute the full convolution
-                let mut pk_body_convolved = vec![Scalar::ZERO; lwe_dimension.0];
-
-                slice_semi_reverse_negacyclic_convolution(
-                    output_mask.as_mut(),
-                    pk_mask.as_ref(),
-                    binary_random_slice,
-                );
+                // output_body_chunk may not be able to fit the full convolution result so we
+                // create a temp buffer to compute the full convolution
+                let mut pk_body_convolved = vec![Scalar::ZERO; max_ciphertext_per_bin];
 
-                // Fill the temp buffer with b convolved with r
-                slice_semi_reverse_negacyclic_convolution(
-                    pk_body_convolved.as_mut_slice(),
-                    pk_body.as_ref(),
-                    binary_random_slice,
+                rayon::join(
+                    || {
+                        slice_semi_reverse_negacyclic_convolution(
+                            output_mask.as_mut(),
+                            pk_mask.as_ref(),
+                            binary_random_slice,
+                        );
+                    },
+                    || {
+                        // // Fill the temp buffer with b convolved with r
+                        slice_semi_reverse_negacyclic_convolution(
+                            pk_body_convolved.as_mut_slice(),
+                            pk_body.as_ref(),
+                            binary_random_slice,
+                        );
+                    },
                 );
 
-                // Noise from Chi_1 for the mask part of the encryption
-                loop_generator
-                    .unsigned_integer_slice_wrapping_add_random_noise_from_distribution_assign(
-                        output_mask.as_mut(),
-                        mask_noise_distribution,
-                    );
+                let zk_body_convolved = polymul_rev(pk_body.as_ref(), binary_random_slice);
 
-                // Fill the body chunk afterwards manually as it most likely will be smaller than
-                // the full convolution result. rev(b convolved r) + Delta * m + e2
-                // taking noise from Chi_2 for the body part of the encryption
-                // The reverse is to make the first element product match the single ciphertext case
+                assert_eq!(&pk_body_convolved, &zk_body_convolved);
+
+                slice_wrapping_add_assign(output_mask.as_mut(), mask_noise);
+
+                // Fill the body chunk afterward manually as it most likely will be smaller than
+                // the full convolution result. b convolved r + Delta * m + e2
```

Comment:
careful with the comment and the reverse we did, as we can see above I added a rev at some point in the comment to match the zk order, the comment seems to be missing this ?